### ISSUE TITLE
Wrong positions of entities and static meshes

### ISSUE THREAD

**Issue by vvs-** [opentomb/OpenTomb#145]
Some entities and static meshes have incorrect positions which are slightly off.

**Comment by stohrendorf** [comment]
How much is "slightly"? Can you make a screenshot for example?

**Comment by vvs-** [comment]
![screenshot from 2015-07-10 15 55 18](https://cloud.githubusercontent.com/assets/5171934/8619106/af261b08-271c-11e5-94f6-1446278a3698.png)
![screenshot from 2015-07-10 15 57 17](https://cloud.githubusercontent.com/assets/5171934/8619114/c5622f24-271c-11e5-8904-8b435bb28272.png)

**Comment by Lwmte** [comment]
By the way, what's up with FPS label colour? Did you changed it in script, or fonts are all pitch black?

**Comment by vvs-** [comment]
Right, that's why debug info isn't working. I didn't notice that. Thanks, Lwmte!

Font colors on console are correct.

**Comment by stohrendorf** [comment]
These are probably the entity angles, which are held internally as degrees with switched axes, which in turn makes it a bit difficult to find the real cause.

And by the way: What machine specs do you have @vvs- ? Because I have more than 1000 FPS in debug mode, and hopefully there isn't a deeply hidden physics bug eating up the frames...

**Comment by vvs-** [comment]
No, there are no problems with performance. By default the graphics driver is always use vblank synchronizing, so I always have 60 FPS. I tried to switch it off once, but it hasn't satisfied me because it was too jerky.

**Comment by vvs-** [comment]
The entity has a correct position, I compared it with old engine. It's just the rendering that's wrong.

**Comment by vvs-** [comment]
That's the reason for broken font colors:
```
diff --git a/src/gui.cpp b/src/gui.cpp
index 899c737..e8cfbde 100644
--- a/src/gui.cpp
+++ b/src/gui.cpp
@@ -496,7 +496,7 @@ void Gui_RenderStringLine(gui_text_line_p l)
     }
 
     std::copy(style->real_color+0, style->real_color+4, gl_font->gl_font_color);
-    // glf_render_str(gl_font, real_x, real_y, l->text);
+    glf_render_str(gl_font, real_x, real_y, l->text);^M
 }
 
 void Gui_RenderStrings()
```

**Comment by stohrendorf** [comment]
Well, the entity angles are pretty ugly, as they are in degrees and in the TR-local coordinate system (I think). When I fix the switch's rotation, another rotation goes nuts. Unfortunately there are 6 ways to permute the axes, 8 combinations in negating the axes, and another 6 possibilities for applying the rotation order, giving 287 ways to do things wrong. I have no idea, honestly. And I also don't know why the textures are wrong, but I strongly suspect the coordinates are wrong.

**Comment by vvs-** [comment]
> And I also don't know why the textures are wrong, but I strongly suspect the coordinates are wrong.

It's not a suspicion - it's a fact. I already tested it before and wrote about that. The only thing that works right is the old texture atlas, that's why I reverted to it in my branch.

**Comment by vvs-** [comment]
> giving 287 ways to do things wrong. I have no idea, honestly.

That's simply because it's not possible to develop that engine single-handedly. You need help from others who familiar with the code. I already said that this code base is too complex for a single person. It'd take ages to find all bugs without intimate knowledge.

**Comment by vvs-** [comment]
Guys, we aren't going anywhere until you all help stohrendorf to fix all bugs. If you feel that it can't be done then there is no point to fish pie in the sky.

**Comment by Lwmte** [comment]
I'm sorry, but we **need** to keep entity rotations in angles for the sake of backwards compatibility with old TRs, as well as because I use **rotateEntity** script command to continously rotate specific entities (for ex., Thames Wharf machine, teeth spikes, etc.). And if it worked before code change, it means there *was* a correct way to do this?

I'll gladly help, but at first I need to understand which branch I should use. Currently there is **glew**, and **luastate** branches, not mentioning **master**, cause it was updated ages ago. Maybe we merge glew and luastate, and then merge it into master? I won't do this, as I'm afraid to mess everything up even more.

**Comment by Lwmte** [comment]
P.S.: If there is a way to get rid of Euler rotations and replace them with something else, then I have no objections, but all scripts should be updated to keep things working.

**Comment by vvs-** [comment]
The `glew` branch is just a provisional thing in case we *want* to merge `luastate` piece by piece. It supposed to be self-contained without other changes. I already proposed to stop introducing any new features or API changes in `luastate` and declare it for bug fixing only.

**Comment by vvs-** [comment]
> If there is a way to get rid of Euler rotations and replace them with something else, then I have no objections

That should be done only *after* everything already fixed. We can't keep fixing bugs and introduce new things at once. It'll just not work.

**Comment by vvs-** [comment]
In my opinion `luastate` now is as good as `master`. So, it's easier to throw away that `glew` branch (which only contains selected commits from `luastate` anyway) and start fixing bugs directly in `luastate`

**Comment by stohrendorf** [comment]
Who wants to merge? :smiling_imp:

**Comment by Lwmte** [comment]
Done! Let's hope I didn't mess things up.

**Comment by vvs-** [comment]
In fact you could just continue to work in `luastate` without merging it. But I don't see any harm in that either.

**Comment by stohrendorf** [comment]
Seems to have worked quite well. At least the bugs I remember are still the same...

**Comment by vvs-** [comment]
Yes, and the same commits wrongly closed the same issues :smiling_imp:

**Comment by vvs-** [comment]
> When I fix the switch's rotation, another rotation goes nuts.

Yes, that's for sure. There are now many more incorrect angles.

**Comment by vvs-** [comment]
Entities have correct position and boundary boxes and just rendered with offset. But the static meshes really have incorrect positions and boundary boxes.

**Comment by vvs-** [comment]
Related to #140.

**Comment by vvs-** [comment]
I thought that it was fixed but now I've encountered with another case. In *The Great Wall* entity 93 rendered at incorrect position.

**Comment by vvs-** [comment]
Here is a screenshot:
![screenshot from 2015-07-21 21 49 06](https://cloud.githubusercontent.com/assets/5171934/8809596/5be0bd02-2ff2-11e5-9ef6-d71135672188.png)

**Comment by Lwmte** [comment]
Fixed in a7c5fe2

**Comment by vvs-** [comment]
Still doesn't fixed. Here's screenshot:
![screenshot from 2015-08-04 22 31 19](https://cloud.githubusercontent.com/assets/5171934/9070605/4cd14b1c-3af9-11e5-8207-bf6aaffd2175.png)
That's before the refactoring:
![screenshot from 2015-08-04 22 34 21](https://cloud.githubusercontent.com/assets/5171934/9070592/3b72814c-3af9-11e5-8218-1561f218ceef.png)

**Comment by vvs-** [comment]
Seems that it's a different bug. The initial position is correct but it was changed at some point.

**Comment by stohrendorf** [comment]
I have been there after the refactoring, and I haven't observed a misplacement. Maybe it's related to animation/scripts, but I don't think it's a wrong placement at first. You can see that it's only a misplacement, not a misrotation.

**Comment by vvs-** [comment]
Yes, It just looked similar. This is something rare which I can't reproduce at will. The initial position is above that point. In normal circumstances the handle is moved forward along the line when the level is started. But seems that sometimes it just drops below. May be it's collision related.

**Comment by vvs-** [comment]
I can reproduce it but not reliably if I put camera there before starting the level. The animation is wrong. The handle is moved forward and then briefly appears behind that point again. Sometimes it's stuck there.

**Comment by Lwmte** [comment]
Is it happening only with death slide entity? If so, I think we won't bother about it, since it needs to be scripted to position correctly at *any* given time.

**Comment by vvs-** [comment]
I have no idea. I saw it just now and only with the zipline handle, but I can't say what's happening and couldn't it affect other entities as well.

**Comment by vvs-** [comment]
I suspect that it can happen with any entities which must be moved during activation. But I can't remember anything else besides zipline.
